clients/client-coinbase/src/utils.ts

```diff
@@ -0,0 +1,1080 @@
+			? PROSPER_STAKING_CONTRACT_ADDRESS
+			: RSP_STAKING_CONTRACT_ADDRESS,
+	);
+	const totalSupply = BigInt(await contract.totalSupply());
+
+	// if (totalSupply === BigInt(0)) {
+	//   totalSupply = BigInt(9999999999999999999000);
+	// }
+	console.log("Total Supply: " + totalSupply);
+
+	// Calculate total staked USD using number arithmetic
+	const totalStakedUSD = Number(totalSupply) * tokenPrice;
+
+	console.log("Total Staked USD" + totalStakedUSD);
+	console.log("Token Price " + tokenPrice);
+
+	return totalStakedUSD.toFixed(4); // Return as a formatted string
+};
+
+export const getUserStaked = async (
+	userWallet: string,
+	token: "PROSPER" | "RSP",
+) => {
+	const contract = getContract(
+		token === "PROSPER"
+			? PROSPER_STAKING_CONTRACT_ADDRESS
+			: RSP_STAKING_CONTRACT_ADDRESS,
+	);
+	const balance = await contract.balanceOf(userWallet);
+	return formatUnits(balance, 18);
+};
+
+export const getAvailableRewards = async (
+	userWallet: string,
+	token: "PROSPER" | "RSP",
+) => {
+	const contract = getContract(
+		token === "PROSPER"
+			? PROSPER_STAKING_CONTRACT_ADDRESS
+			: RSP_STAKING_CONTRACT_ADDRESS,
+	);
+	const rewards = await contract.earned(userWallet, 0);
+	return formatUnits(rewards, 18);
+};
+
+export const claimReward = async (signer: Signer, token: "PROSPER" | "RSP") => {
+	const contract = getContract(
+		token === "PROSPER"
+			? PROSPER_STAKING_CONTRACT_ADDRESS
+			: RSP_STAKING_CONTRACT_ADDRESS,
+		signer,
+	);
+	const tx = await contract.getReward();
+	await tx.wait();
+};
+
+export const unstake = async (signer: Signer, token: "PROSPER" | "RSP") => {
+	const contract = getContract(
+		token === "PROSPER"
+			? PROSPER_STAKING_CONTRACT_ADDRESS
+			: RSP_STAKING_CONTRACT_ADDRESS,
+		signer,
+	);
+	const tx = await contract.exit();
+	await tx.wait();
+};
+
+export const stake = async (
+	signer: Signer,
+	amount: string,
+	token: "PROSPER" | "RSP",
+) => {
+	const contract = getContract(
+		token === "PROSPER"
+			? PROSPER_STAKING_CONTRACT_ADDRESS
+			: RSP_STAKING_CONTRACT_ADDRESS,
+		signer,
+	);
+	const tx = await contract.stake(amount);
+	await tx.wait();
+};
+
+export const getAvailableToken = async (
+	userWallet: string,
+	token: "PROSPER" | "RSP",
+) => {
+	const rpcUrl =
+		"https://base-mainnet.infura.io/v3/f2ace3972c5d4c75bfd063d4016d423a";
+	const provider = new JsonRpcProvider(rpcUrl);
+	const prosperContract = new Contract(
+		token === "PROSPER" ? PROSPER_STAKING_TOKEN : RSP_STAKING_TOKEN,
+		ABI,
+		provider,
+	); // token being staked
+	const balance = await prosperContract.balanceOf(userWallet);
+	// console.log("AAAAA alance: " + balance, userWallet, token)
+	return formatUnits(balance, 18);
+};
+
+export const getRewardPerToken = async (token: "PROSPER" | "RSP") => {
+	const contract = getContract(
+		token === "PROSPER"
+			? PROSPER_STAKING_CONTRACT_ADDRESS
+			: RSP_STAKING_CONTRACT_ADDRESS,
+	);
+	const rewardPerToken = await contract.rewardPerToken(0); // Assuming pool ID 0
```

Comment:
Why we're assuming pool ID 0 ?

Reply:
Not sure but know it works grabbing from working repo wrote